compiler-core/src/type_.rs

```diff
@@ -63,6 +63,7 @@ pub enum Type {
         module: EcoString,
         name: EcoString,
         args: Vec<Arc<Type>>,
+        constructor_index: Option<u16>,
```

Comment:
What's this field? It's not clear to me how a type has a constructor index.

Reply:
This represents which variant a certain value has when the variant is known, so for example:
```gleam
type Wibble { Wibble Wobble }

fn wibble(w: Wibble) {
  //      ^ We have no idea which variant this is, so constructor_index is None
  case w {
    Wibble as w -> todo
    //        ^ We know that this is the Wibble variant, so constructor_index would be Some(0)
    Wobble as w -> todo
  }
}
```

This perhaps isn't the best way to implement this, but it's the way I came up with that made fairly minimal changes to the way that the type system is represented internally. If you have another idea, I'm open to suggestions

Reply:
I don't have any particular suggestions for the design, I think this is quite clever.

I'm wondering if there's a clearer name that we can use here as it wasn't obvious to me what this means. At the least it needs a good deal of documentation.